KarhooUISDK/UIComponents/LegalNotice/KarhooLegalNoticeView.swift

```diff
@@ -0,0 +1,135 @@
+    // MARK: - Setup
+    private func setUpView() {
+        accessibilityIdentifier = KHLegalNoticeViewID.view
+        translatesAutoresizingMaskIntoConstraints = false
+        
+        attributedLabel.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(self.singleTap(_:))))
+        
+        addSubview(legalNoticeButton)
+        addSubview(attributedLabel)
+        hideLegalNoticePressed()
+    }
+    
+    override func updateConstraints() {
+        if !didSetUpConstraints {
+            legalNoticeButton.anchor(
+                top: topAnchor,
+                leading: leadingAnchor,
+                bottom: attributedLabel.topAnchor,
+                paddingLeft: 8.0,
+                paddingRight: 8.0)
+            
+            attributedLabel.anchor(
+                top: legalNoticeButton.bottomAnchor,
+                leading: leadingAnchor,
+                bottom: bottomAnchor,
+                trailing: trailingAnchor,
+                paddingLeft: 8.0,
+                paddingBottom: 16,
+                paddingRight: 8.0)
+            
+            didSetUpConstraints.toggle()
+        }
+        super.updateConstraints()
+    }
+    
+    @objc func singleTap(_ tap : UITapGestureRecognizer) {
+           let attributedText = NSMutableAttributedString(attributedString: self.attributedLabel.attributedText!)
+        attributedText.addAttributes([NSAttributedString.Key.font: self.attributedLabel.font as UIFont], range: NSMakeRange(0, (self.attributedLabel.attributedText?.string.count)!))
+
+           // Create instances of NSLayoutManager, NSTextContainer and NSTextStorage
+           let layoutManager = NSLayoutManager()
+           let textContainer = NSTextContainer(size: CGSize(width: (self.attributedLabel.frame.width), height: (self.attributedLabel.frame.height)+100))
+           let textStorage = NSTextStorage(attributedString: attributedText)
+
+           // Configure layoutManager and textStorage
+           layoutManager.addTextContainer(textContainer)
+           textStorage.addLayoutManager(layoutManager)
+
+           // Configure textContainer
+           textContainer.lineFragmentPadding = 0.0
+           textContainer.lineBreakMode = self.attributedLabel.lineBreakMode
+           textContainer.maximumNumberOfLines = self.attributedLabel.numberOfLines
+           let labelSize = self.attributedLabel.bounds.size
+           textContainer.size = labelSize
+
+           let tapLocation = tap.location(in: self.attributedLabel)
+
+           // get the index of character where user tapped
+           let index = layoutManager.characterIndex(for: tapLocation, in: textContainer, fractionOfDistanceBetweenInsertionPoints: nil)
+
+           if index > (self.attributedLabel.text?.count)! { return }
+           var range : NSRange = NSRange()
+        if let _ = self.attributedLabel.attributedText?.attribute(NSAttributedString.Key(rawValue: "link"), at: index, effectiveRange: &range) as? String {
```

Comment:
1. Sooo complex statement. Break to multiline & use some scope constants, so it is easier for reader to follow
2. Consider following `early return` approach (https://google.github.io/swift/#guards-for-early-exits)